Added ImpactRotation and ImpactPosition to WarheadArgs

reaperrr opened the pull request:
See commits for more details.

Mostly for downstream projects for now.

I'll provide an example testcase soon (probably tomorrow).

reaperrr committed:
Add ImpactOrientation to WarheadArgs

Allows to pass the horizontal facing/yaw
and vertical angle/pitch of the carrier
projectile to warheads for further use.

Add ImpactPosition to WarheadArgs

InflictDamage doesn't pass the impact pos
directly, and the very point of WarheadArgs
is to avoid adding more and more arguments
to the warhead methods.

reaperrr committed:
Adapt SpreadDamageWH to ImpactOrientation

The ImpactOrientation needs to be computed from
point of impact to target if the target wasn't hit directly.

Also adapted warhead code to use WarheadArgs consistently,
as well as pass HitShape instead of just HitShapeInfo
(both needed for future and/or downstream features).

Mailaender commented:
Both `ImpactPosition` and `ImpactRotation` seem to be unused. Why do you want to set those at time of impact? Wouldn't it make more sense for a projectile to get rotation at launch to match the launcher or rotate it while it is in the air and flying an arch?

reaperrr commented:
Because for warheads, only the projectile position and rotation at the moment of impact matters, I don't see the point in setting them earlier or updating on the fly (that would just cost performance).

I'm using them downstream for my directional armor implementation, which I might submit upstream after this (not sure if my approach would get conceptual approval, but maybe it will).

reaperrr commented:
Note: I'll hold off rebasing this until #18023 is merged (which hopefully won't take too much longer).

pchote commented on OpenRA.Game/GameRules/WeaponInfo.cs:
```diff
@@ -38,13 +38,16 @@ public class WarheadArgs
 		public WeaponInfo Weapon;
 		public int[] DamageModifiers = { };
 		public WPos? Source;
+		public WRot ImpactRotation;
```
Would `ImpactOrientation` still make sense here? This would make it consistent with the actor orientation naming.

pchote reviewed (commented):
LGTM, just one naming comment/question.

I haven't tried testing this yet because this doesn't touch anything used in the upstream mods.

reaperrr commented:
Rebased and updated.

Note: Due to rebase conflicts caused by the recent Facing int->WAngle changes, I decided to merge the first 2 commits to save me from more rebase conflict hell.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c04fd6c08509a8386fcf23380740bdabb1a5f6eb...7e2eaa697bd84c1bb670606e1fcb0ac030923f5d
